Copy checkstyle config to allow running from a jar

jaymode opened the pull request:
The checkstyle configuration files were being accessed as resources within the project and
being converted from a URL to a File by gradle. This works when the build tools project is being
referenced as a project. However, when using the published jar the URL points to a resource
in the jar file that cannot be converted to a File object and caused the build to fail.

This change copies the files into a `checkstyle` directory in the project build folder and always uses
File objects pointing to the copied files.

jaymode committed:
Copy checkstyle config to allow running from a jar

The checkstyle configuration files were being accessed as resources within the project and
being converted from a URL to a File by gradle. This works when the build tools project is being
referenced as a local project. However, when using the published jar the URL points to a resource
in the jar file, that URL cannot be converted to a File object and causes the build to fail.

This change copies the files into a `checkstyle` directory in the project build folder and always uses
File objects pointing to the copied files.

jaymode commented:
@rjernst can you take a look?

ywelsch commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/PrecommitTasks.groovy:
```diff
@@ -87,16 +87,30 @@ class PrecommitTasks {
         project.checkstyle {
-            config = project.resources.text.fromFile(
-                PrecommitTasks.getResource('/checkstyle.xml'), 'UTF-8')
+            config = project.resources.text.fromFile(new File(checkStyleDir, "checkstyle.xml"), 'UTF-8')
```
instead of creating a file copying the content etc, you can probably make this a single line change:
`config = project.resources.text.fromString(new String(PrecommitTasks.getResource("/checkstyle_suppressions.xml").openStream().getBytes()))`

jaymode replied:
nice! I pushed a commit that does this. A separate task is still required (explained in a code comment)

nik9000 commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/PrecommitTasks.groovy:
```diff
@@ -87,26 +87,39 @@ class PrecommitTasks {
+        // We need access to the checkstyle configuration and this needs to be done after configuration otherwise running a command
+        // like 'gradle clean precommit' will fail since the suppressions needs to be a file and if it is created during configuration
+        // then it will be deleted during clean
+        Task configCheckstyle = project.tasks.create('configCheckstyle').doLast {
```
Usually we put `doLast` on a new line I think.

rjernst replied:
And don't do it on the same line as creating.

```
Task configCheckstyle = project.tasks.create('configCheckstyle')
configCheckstyle.doLast {
...
```

nik9000 commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/PrecommitTasks.groovy:
```diff
@@ -87,26 +87,39 @@ class PrecommitTasks {
+        // then it will be deleted during clean
+        Task configCheckstyle = project.tasks.create('configCheckstyle').doLast {
+            File checkstyleSuppressions = project.resources.text.fromString(
+                    new String(PrecommitTasks.getResource("/checkstyle_suppressions.xml").openStream().getBytes())).asFile()
```
I wonder about this because I *think* we read stuff from the resources like this all the time and I think it works with `clean` because these reseources are part of buildsrc. I thought.

nik9000 replied:
Oh! I see - this probably is/was the only place that did the asFile thing.

nik9000 replied:
OK -I think something like this is fine but please be careful to keep the UP-TO-DATE check working for checkstyle - if you run it twice the second time gradle should just skip it as UP-TO-DATE. That will help to keep development times shorter.

rjernst replied:
This is not the right way to read a file, it will never close the stream. It also is using the default file encoding. I would just stick with the original, all this is going to do is have gradle copy it to a file instead of you.

jaymode commented:
@nik9000 @rjernst I reverted the change to how the files are copied and added a new commit that sets inputs/outputs so that the UP-TO-DATE checks work

nik9000 commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/PrecommitTasks.groovy:
```diff
@@ -87,16 +87,43 @@ class PrecommitTasks {
+
+        // configure inputs and outputs so up to date works properly
+        copyCheckstyleConf.outputs.files(checkstyleSuppressions, checkstyleConf)
+        if ("jar".equals(checkstyleConfUrl.getProtocol())) {
```
Oh! I'm not surprised this is required but it is silly looking. Good catch though.

nik9000 commented:
I'm happy with it. @rjernst knows gradle much better than I do though.

rjernst commented:
lgtm
